_source/_posts/2022-06-28-add-auth-to-any-app-with-oauth2-proxy.adoc

```diff
@@ -0,0 +1,583 @@
+      OAUTH2_PROXY_OIDC_EXTRA_AUDIENCES: api://default # <.>
+      OAUTH2_PROXY_OIDC_EMAIL_CLAIM: sub # <.>
+
+      OAUTH2_PROXY_SET_XAUTHREQUEST: true # <.>
+      OAUTH2_PROXY_PASS_ACCESS_TOKEN: true # <.>
+
+      OAUTH2_PROXY_SESSION_STORE_TYPE: redis # <.>
+      OAUTH2_PROXY_REDIS_CONNECTION_URL: redis://redis
+
+      OAUTH2_PROXY_COOKIE_REFRESH: 30m # <.>
+      OAUTH2_PROXY_COOKIE_NAME: SESSION # <.>
+      OAUTH2_PROXY_COOKIE_SECRET: ${OAUTH2_PROXY_COOKIE_SECRET} # <.>
+
+  nginx:
+    image: nginx:1.21.6-alpine
+    depends_on:
+      - oauth2-proxy
+      - web-app
+    volumes:
+      - ./nginx-default.conf.template:/etc/nginx/templates/default.conf.template
+    ports:
+      - 80:80
+
+  redis:
+    image: redis:7.0.2-alpine3.16
+    volumes:
+      - cache:/data
+
+volumes:
+  cache:
+    driver: local
+----
+
+<.> Allow all email addresses; the IdP will manage which users have access.
+<.> For single IdP use cases, skip the intermediate login page.
+<.> oauth2-proxy defaults to `https`, this example uses `http` on localhost.
+<.> The issuer, client id, and secret will be loaded from the `.env` file.
+<.> Allow processing of JWT bearer tokens for API clients.
+<.> Configure an extra "allowed" audience, in addition to the "client ID".
+<.> Use the `sub` claim from JWT access tokens as the email address.
+<.> Add user information headers to the proxied web-app request.
+<.> Optional, pass the access token to the proxied web-app request.
+<.> Use Redis for session management.
+<.> Refresh cookies every 30 minutes.
+<.> Set the session cookie name to `SESSION`.
+<.> Configure the encryption key (loaded from the `.env` file).
+====
+
+Restart the services and access the application through your browser: `http://localhost/echo`. Try again using HTTPie:
+
+[source,bash]
+----
+http localhost/echo "Authorization: Bearer ${TOKEN}"
+----
+
+Both requests should show similar information!
+
+Without any code changes, the "echo" web application is now secured with OIDC / OAuth 2.0!
+
+== Is OAuth2 Proxy right for your application?
+
+This post taught you how to secure an existing application with OAuth 2.0 without any code changes! OAuth2 Proxy isn't just for legacy applications; it's also commonly used in a sidecar pattern with Kubernetes, allowing you to keep authorization concerns separate from your application.
+
+It's not perfect, though; there are a few shortcomings:
+
+- SSO logout (link:/blog/2020/03/27/spring-oidc-logout-options[OIDC RP-Initated Logout]) isn't supported yet.
+- Requires working around a logout route that accepts `GET` requests and doesn't require a CSRF token.
+- API use cases only support JWTs access tokens; if your application needs to link:/blog/2020/08/07/spring-boot-remote-vs-local-tokens[remotly validate tokens] you would need another solution. It may be easier to configure https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html[Nginx's JWT module] instead.
```

Comment:
```suggestion
- API use cases support JWTs only as access tokens; if your application needs to link:/blog/2020/08/07/spring-boot-remote-vs-local-tokens[remotely validate tokens] you would need another solution. It may be easier to configure https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html[Nginx's JWT module] instead.
```
typo, grammar

Reply:
```suggestion
- API use cases support JWT only access tokens; if your application needs to link:/blog/2020/08/07/spring-boot-remote-vs-local-tokens[remotely validate tokens] you would need another solution. It may be easier to configure https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html[Nginx's JWT module] instead.
```